analysis/topology/topological_space.lean

```diff
@@ -718,8 +1111,41 @@ let ⟨t, ht₁, ht₂, ht₃⟩ := this in
   subset.trans (compl_subset_comm.1 ht₂) h₁,
   is_closed_compl_iff.mpr ht₁⟩
 
+variable (α)
+instance regular_space.t2_space [regular_space α] : t2_space α :=
+{ t2 := λ x y hxy,
+    let ⟨s, h1, h2, h3⟩ := regular_space.regular is_closed_singleton
+          (mt mem_singleton_iff.1 hxy),
+        ⟨t, h4, u, h5, h6⟩ := empty_in_sets_eq_bot.2 h3,
+        ⟨v, h7, h8, h9⟩ := mem_nhds_sets_iff.1 h4 in
+    ⟨v, s, h8, h1, h9, singleton_subset_iff.1 h2,
+      eq_empty_of_subset_empty $ λ z ⟨h10, h11⟩, h6 ⟨h7 h10, h5 h11⟩⟩ }
+
 end regularity
 
+section normality
+
+/-- A T₄ space, also known as a normal space (although this condition sometimes
+  omits T₂), is one in which for every disjoint closed `C` and `D`, there exist
+  disjoint open sets containing `C` and `D` respectively. -/
+class normal_space (α : Type u) [topological_space α] extends t1_space α : Prop :=
+(normal : ∀ s t : set α, is_closed s → is_closed t → disjoint s t →
+  ∃ u v, is_open u ∧ is_open v ∧ s ⊆ u ∧ t ⊆ v ∧ disjoint u v)
+
+theorem normal_separation [normal_space α] (s t : set α)
+  (H1 : is_closed s) (H2 : is_closed t) (H3 : disjoint s t) :
+  ∃ u v, is_open u ∧ is_open v ∧ s ⊆ u ∧ t ⊆ v ∧ disjoint u v :=
+normal_space.normal s t H1 H2 H3
+
+variable (α)
+instance normal_space.regular_space [normal_space α] : regular_space α :=
+{ regular := λ s x h1 h2, let ⟨u, v, h1, h2, h3, h4, h5⟩ := normal_separation s {x} h1 is_closed_singleton
```

Comment:
yuck. It should be no problem to come up with some more helpful names than these.

Reply:
Well I don't see how `normal_space.regular_space` is not informative.

Reply:
I would guess that Reid is objecting to all the intermal variables `u, v, h1, h2, h3, h4, ...`, not the name of the instance.

Reply:
I agree with Reid on this. Reading your vector space dimension theorem was very unpleasant when you'd got up to `h17` and there were defined functions, hypotheses and assumptions of all sorts under the same naming scheme.

If you want a naming algorithm: I recommend using 2-3 letter abbreviations of some major distinguishing characteristic of the type of the assumption for hypotheses, and one letter names for the "objects" that appear in other types, not including subscripts which can be added for very similar or logically related values.